### PR TITLE
Valid blocks

### LINKED ISSUE #23557: Allow the user to specify not (yet) existing subdomain IDs as valid
## Reason
With dynamic subdomain changes needed for AM or element death simulations, we often run into the limitation that the set of valid subdomains is the set of subdomains existing on the initial mesh. However at t=0 no element might be in the deleted subdomain. 

We currently circumvent this by having a free floating element unconnected to the mesh, with the needed subdomain id.

## Design

I propose a new parameter

```
Mesh/add_subdomain_id = 9999
```

that lets the user specify a list of IDs that are assumed to be valid. That will permit the user to set up block restrictions for this yet non-existing subdomain in their inputs.

## Impact
Simplified capability.

Ping @bwspenc

### PR DESCRIPTION
**Pull request by dschwen**
Closes #23557

### CODE REVIEW

File: framework/src/mesh/MooseMesh.C [hunk 1]
```diff
@@ -147,6 +147,12 @@ MooseMesh::validParams()
                         true,
                         "True to skip uniform refinements when using a pre-split mesh.");
 
+  params.addParam<std::vector<SubdomainID>>(
+      "add_subdomain_ids",
+      "The listed subdoamins will be assumed valid for the mesh. This permits setting up subdomain "
+      "restrictions for subomains containing no elements, for example in simulations, where "
```
Comment (lindsayad) [inline framework/src/mesh/MooseMesh.C:153]:
```suggestion
      "The listed subdomains will be assumed valid for the mesh. This permits setting up subdomain "
      "restrictions for subdomains initially containing no elements, which can occur, for example, in "
      "additive manufacturing simulations which dynamically add and remove elements."
```

File: framework/src/mesh/MooseMesh.C [hunk 2]
```diff
@@ -147,6 +147,12 @@ MooseMesh::validParams()
...
+      "subdomains are dynamically changed.");
```
Comment (lindsayad) [inline framework/src/mesh/MooseMesh.C:154]:
```suggestion
```

File: framework/doc/content/source/mesh/MooseMesh.md [hunk 3]
```diff
@@ -36,6 +36,7 @@ You could also use the [SideSetsFromNodeSetsGenerator.md] class to construct sid
 sets (or the [NodeSetsFromSideSetsGenerator.md] for vice versa). These object allows you to manually
 construct the relevant side sets and node sets earlier in the mesh generation process.
 
+To declare subdomain IDs which are not found on the mesh right after the mesh is setup, a list of IDs can be specified using the `Mesh/add_subdomain_ids` parameter. This permits setting up block restricted objects that will only act on subdomains created during the simulation (e.g. using the[CoupledVarThresholdElementSubdomainModifier.md]).
```
Comment (lindsayad) [inline framework/doc/content/source/mesh/MooseMesh.md:39]:
```suggestion
To declare subdomain IDs which are not found on the mesh right after the mesh is setup, a list of IDs can be 
specified using the `Mesh/add_subdomain_ids` parameter. This permits setting up block restricted objects that
will only act on subdomains created during the simulation (e.g. using the [CoupledVarThresholdElementSubdomainModifier.md]).
```

File: test/tests/mesh/add_subdomain_ids/add_subdomain_ids.i [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    value = 1
+  []
+[]
+
+[Executioner]
+  type = Steady
+
+  # Preconditioned JFNK (default)
+  solve_type = 'PJFNK'
+
```
Comment (lindsayad) [inline test/tests/mesh/add_subdomain_ids/add_subdomain_ids.i:49]:
```suggestion
```

Comment (lindsayad) [inline test/tests/mesh/add_subdomain_ids/add_subdomain_ids.i:49]:
```suggestion
```

**Review by lindsayad — CHANGES_REQUESTED** [review]

File: test/tests/mesh/add_subdomain_ids/tests [hunk 5]
```diff
@@ -0,0 +1,10 @@
+[Tests]
+  [add_subdomain_ids]
+    type = Exodiff
+    input = add_subdomain_ids.i
+    exodiff = add_subdomain_ids_out.e
+    issues = '#23557'
+    design = MooseMesh.md
+    requirement = 'The system shall allow specifying valid subdomain IDs that do not exist on the initial mesh'
```
Comment (lindsayad) [inline test/tests/mesh/add_subdomain_ids/tests:8]:
```suggestion
    requirement = 'The system shall allow specifying valid subdomain IDs that do not exist on the initial mesh.'
```
My SQA sources tell me we're supposed to have these be complete sentences

**Review by lindsayad — CHANGES_REQUESTED** [review]

**Comment by dschwen** [comment]
I'll squash. GitHub didn't let me apply all suggestions at the same time, because some applied to the same line.

**Comment by lindsayad** [comment]
GitHub was really throwing me off. It made a suggestion on the 4 lines appear as if I'd only suggested on the top 3 lines which was why I added the single line deletion ... it seems what GitHub was showing me for my suggestion didn't match what it applied in practice :-/

**Review by lindsayad — APPROVED** [review]

**Pull request merged**